File: 3sum/Yiseull.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        answer = set()

        nums = sorted(nums)

        n = len(nums)
        for i in range(n - 2):
            if nums[i] > 0:
                break

            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, n - 1
            while left < right:
                threeSum = nums[i] + nums[left] + nums[right]
                if threeSum < 0:
                    left += 1
                elif threeSum == 0:
                    answer.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1
                else:
                    right -= 1

        return list(answer)

File: search-in-rotated-sorted-array/Yiseull.java
class Solution {
    // 공간복잡도: O(1), 시간복잡도: O(logn)
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left < right) {
            int mid = (left + right) / 2;

            // 오른쪽 정렬
            if (nums[mid] < nums[right]) {
                if (nums[mid] <= target && target <= nums[right]) {
                    if (nums[mid] == target) return mid;
                    left = mid + 1;
                }
                else right = mid - 1;
            // 왼쪽 정렬
            } else {
                if (nums[left] <= target && target <= nums[mid]) {
                    if (nums[mid] == target) return mid;
                    right = mid - 1;
                }
                else left = mid + 1;
            }
        }

        return nums[left] == target ? left : -1;
    }
}
